src/Std/Sat/AIG/Basic.lean

```diff
@@ -0,0 +1,475 @@
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    Ref aig2 :=
+  { ref with hgate := by have := ref.hgate; omega }
+
+structure BinaryInput (aig : AIG α) where
+  lhs : Ref aig
+  rhs : Ref aig
+
+@[inline]
+def BinaryInput.cast {aig1 aig2 : AIG α} (input : BinaryInput aig1)
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    BinaryInput aig2 :=
+  { input with lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+structure TernaryInput (aig : AIG α) where
+  discr : Ref aig
+  lhs : Ref aig
+  rhs : Ref aig
+
+@[inline]
+def TernaryInput.cast {aig1 aig2 : AIG α} (input : TernaryInput aig1)
+    (h : aig1.decls.size ≤ aig2.decls.size) :
+    TernaryInput aig2 :=
+  { input with discr := input.discr.cast h, lhs := input.lhs.cast h, rhs := input.rhs.cast h }
+
+/--
+An entrypoint into an `AIG`. This can be used to evaluate a circuit, starting at a certain node,
+with `AIG.denote` or to construct bigger circuits on top of this specific node.
+-/
+structure Entrypoint (α : Type) [DecidableEq α] [Hashable α] where
+  /--
+  The AIG that we are in.
+  -/
+  aig : AIG α
+  /--
+  The reference to the node in `aig` that this `Entrypoint` targets.
+  -/
+  ref : Ref aig
+
+/--
+Transform an `Entrypoint` into a graphviz string. Useful for debugging purposes.
+-/
+def toGraphviz {α : Type} [DecidableEq α] [ToString α] [Hashable α] (entry : Entrypoint α)
+    : String :=
+  let ⟨⟨decls, _, hinv⟩, ⟨idx, h⟩⟩ := entry
+  let (dag, s) := go "" decls hinv idx h |>.run .empty
+  let nodes := s.fold (fun x y ↦ x ++ toGraphvizString decls y) ""
+  "Digraph AIG {" ++ nodes ++ dag ++ "}"
+where
+  go {α : Type} [DecidableEq α] [ToString α] [Hashable α] (acc : String) (decls : Array (Decl α))
+      (hinv : IsDag α decls) (idx : Nat) (hidx : idx < decls.size)
+        : StateM (HashSet (Fin decls.size)) String := do
+    let fidx : Fin decls.size := Fin.mk idx hidx
+    if (← get).contains fidx then
+      return acc
+    modify (fun s ↦ s.insert fidx)
+    match elem : decls[idx] with
+    | Decl.const _ => return acc
+    | Decl.atom _ => return acc
+    | Decl.gate lidx ridx linv rinv =>
+      let curr := s!"{idx} -> {lidx}{invEdgeStyle linv}; {idx} -> {ridx}{invEdgeStyle rinv};"
+      let hlr := hinv idx lidx ridx linv rinv hidx elem
+      let laig ← go (acc ++ curr) decls hinv lidx (by omega)
+      go laig decls hinv ridx (by omega)
+  invEdgeStyle (isInv : Bool) : String :=
+    if isInv then " [color=red]" else " [color=blue]"
+  toGraphvizString {α : Type} [DecidableEq α] [ToString α] [Hashable α] (decls : Array (Decl α))
+      (idx : Fin decls.size) : String :=
+    match decls[idx] with
+    | Decl.const b => s!"{idx} [label=\"{b}\", shape=box];"
+    | Decl.atom i => s!"{idx} [label=\"{i}\", shape=doublecircle];"
+    | Decl.gate _ _ _ _ => s!"{idx} [label=\"{idx} ∧\",shape=trapezium];"
+
+/--
+A stream of references into `aig`. This is the `AIG` analog of `BitVec`.
+-/
+structure RefStream (aig : AIG α) (length : Nat) where
```

Comment:
If I had `Array Ref` I would have to iterate over the thing every time that the underlying AIG changes and apply `Ref.cast`. This would unnecessarily be `O(n)` and we could of course `implementedBy` this away but that would increase the trust required by LeanSAT. With it done like this the underlying AIG can change and we can update the proof for free all the time.

Reply:
I see, thanks for the explanation.